### ISSUE TITLE
[MLv2] Implement/expose functions for available columns and operators for join conditions

### ISSUE THREAD

**Issue by camsaul** [metabase/metabase#31174]
We need functions to power these parts of the UI, after a Table/etc. is chosen (#28699) and after its `:fields` are selected (#31173).  Note that under some circumstances a join condition is suggested; #31175 exists for tracking that functionality.

There are three parts to building a join condition, the LHS/source column, the operator, and the RHS/target column; you can pick these in any order, but they all have to be specified to have a complete join condition; you must have a complete join condition before you can add a join to a query with MLv2, since MLv2 always maintains a valid query at all times.

### Terminology

So I don't have to explain what this means several times

* _LHS/source column_: the column in the left-hand side of the condition, e.g. the `x` in `JOIN t ON x = y`. `x` either comes from the source Table, or a previous stage of the query, or a previously-joined Table/Model/Saved Question. `x` typically has some sort of FK relationship with the RHS column `y`.
* _RHS/destination/target column_: the column in the right-hand side of the condition e.g. `y` in the example above. `y` is a column in the Table/Model/Saved Question we are joining against. 
* _operator_: `=` in the example above, corresponds to the type of MBQL clause we will be creating. `=` is the default.

##### LHS

<img width="278" alt="image" src="https://github.com/metabase/metabase/assets/1455846/f002ce3b-ca0c-4850-b9f3-ac0076094b42">

###### LHS, including columns from a previous join:

<img width="927" alt="image" src="https://github.com/metabase/metabase/assets/1455846/b8f13029-07d6-4816-a4ed-c851949f8009">

##### Operator

<img width="75" alt="image" src="https://github.com/metabase/metabase/assets/1455846/b0805991-d4b7-4075-89ca-cb0704066bf3">

##### RHS

<img width="395" alt="image" src="https://github.com/metabase/metabase/assets/1455846/ffe5351b-45d0-4fb2-b799-abb188b71724">

### Potential API
For powering the UI to give you the list of LHS columns and RHS columns and operator I think functions like 

```ts
function joinConditionLHSColumns(query: Query,
                                 stageNumber: integer,
                                 joinedTable: TableMetadata,
                                 rhsColumn: ColumnMetadata?): ColumnMetadata[] { ... }

function joinConditionRHSColumns(query: Query,
                                 stageNumber: integer,
                                 joinedTable: TableMetadata,
                                 lhsColumn: ColumnMetadata?): ColumnMetadata[] { ... }

function joinConditionOperators(query: Query,
                                stageNumber: integer,
                                lhsColumn: ColumnMetadata?,
                                rhsColumn: ColumnMetadata?): FilterOperator[] { ... }
```

should do the trick. I think remembering "`LHS` is left and `RHS` is right" is easier than remembering "`source` is left and `target` is right" or something like that.

The expectation is that you'd then pass these column metadatas to something like `groupColumns` and then `displayInfo` to display them, then pick the columns, then pass them to `filterClause` to create the actual opaque MBQL filter clause itself.

#### Temporal columns show bucketing options

There is not currently a way to select a temporal column without bucketing it:

<img width="294" alt="image" src="https://github.com/metabase/metabase/assets/1455846/7fece7f8-2b7f-4977-b620-c991d5c03781">

#### Picking LHS or RHS does not restrict options for the other

LHS and RHS can be picked in either order, and it currently does not affect the results of the other:

<img width="372" alt="image" src="https://github.com/metabase/metabase/assets/1455846/c0e31c94-880d-4eb6-b202-e9ccdab914fc">

This doesn't really make a ton of sense; it makes no sense to let you choose a string column LHS if you already have an integer column as the RHS, but we don't need to rock the boat too much right now. Still, we should leave the options open to have this be smarter about filtering out columns that don't make sense in the future. Thus `LHSColumns(...)` and `RHSColumns(...)` both take an optional value for the other column so we can potentially implement filtering based on their value in the future.

#### Operators are hardcoded 

The list of operators appears to be hardcoded and shows things like `>=` regardless of whether you've already picked an incompatible LHS or RHS column that doesn't support that operator. I think this behavior is broken, but fixing it is not something that needs to happen as part of this issue; we can fix it in the future. To facilitate fixing this in the future I gave `joinConditionOperators()` optional `lhsColumn` and `rhsColumn` parameters so those values can be passed in if we know them.

<img width="156" alt="image" src="https://github.com/metabase/metabase/assets/1455846/a61c5fc2-597f-4580-95d1-24cc9f40c357">

#### Column sort order

I'm not 100% sure what the sort order for the LHS and RHS columns is. It seems that PK and FK columns are shown first, e.g.

1. PK column
2. FK columns
3. ??? Mystery sort order

<img width="245" alt="image" src="https://github.com/metabase/metabase/assets/1455846/8d0e095f-d30d-4d68-8fda-88c796d70304">

<img width="254" alt="image" src="https://github.com/metabase/metabase/assets/1455846/98cd782b-a0d4-4473-892f-d9173a4f7adf">

I don't know how the rest of the columns are sorted. We should try to figure that out and replicate the order the columns get returned in.

**Comment by kulyk** [comment]
@camsaul, should we also include join strategies here?

<img width="445" alt="CleanShot 2023-05-31 at 17 13 06@2x" src="https://github.com/metabase/metabase/assets/17258145/87c763ff-7152-4de9-b96e-e3e165132495">

**Comment by kulyk** [comment]
> **Column sort order**
> I'm not 100% sure what the sort order for the LHS and RHS columns is.

@camsaul I've just checked how it works, and there're no special rules. The picker just follows the order of fields coming from `/api/table/:id/query_metadata` endpoint. But I should say I like the idea to keep PKs and FKs at the top for joins

**Comment by camsaul** [comment]
@kulyk I made join strategies a separate issue, see https://github.com/metabase/metabase/issues/31180. I already have a PR to close that out

**Comment by camsaul** [comment]
One thing worth noting is that the `:fields` of a previous join do not affect the Fields that get shown in the LHS picker. Example:


<img width="933" alt="image" src="https://github.com/metabase/metabase/assets/1455846/49459185-c03f-4d82-b0d1-4c839a3283eb">

<img width="909" alt="image" src="https://github.com/metabase/metabase/assets/1455846/2f32b21e-6909-402e-b358-e68f6c104689">

It shows all the columns regardless as if `:fields` was not specified at all. So this means we should be using `visible columns` or whatever (columns visible within the current stage of the query) rather than the 'exported columns' visible in the results or a potential subsequent stage

**Comment by camsaul** [comment]
Note that implicitly joinable columns are not included here for whatever reason

<img width="308" alt="image" src="https://github.com/metabase/metabase/assets/1455846/310859f7-d9fd-4d0f-bafb-f015abd69326">

**Comment by snoe** [comment]
I think there are existing problems in mlv1 between custom expressions and join conditions. They can become cyclicly dependent. Fields introduced by joins cannot be used in previous joins - even by indirection through custom expressions.


![Image](https://github.com/metabase/metabase/assets/105012/23af76b9-ecfb-46ce-8fd5-69344ef50426)

**Comment by camsaul** [comment]
Implemented by #31226
